pubsub/alicloud/mns/mns.go

```diff
@@ -0,0 +1,303 @@
+		// do nothing here because queue is already created or fetched
+	default:
+		return fmt.Errorf("unsupported MNS mode: %v, should be queue or topic", m.settings.MnsMode)
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	queue := ali_mns.NewMNSQueue(metaData.QueueName, m.client)
+
+	respChan := make(chan ali_mns.MessageReceiveResponse)
+	errChan := make(chan error)
+
+	go m.processMessageLoop(ctx, req.Topic, queue, handler, respChan, errChan)
+	go m.receiveMessageLoop(ctx, queue, respChan, errChan)
+
+	switch m.settings.MnsMode {
+	case MnsModeQueue:
+		if c, ok := m.queueContexts[metaData.QueueName]; ok {
+			c.cancel()
+		}
+		m.queueContexts[metaData.QueueName] = contextWithCancelFunc{ctx, cancelFunc}
+	case MnsModeTopic:
+		if c, ok := m.subscriptionContexts[metaData.SubscriptionName]; ok {
+			c.cancel()
+		}
+		m.subscriptionContexts[metaData.SubscriptionName] = contextWithCancelFunc{ctx, cancelFunc}
+	}
+
+	return nil
+}
+
+func wrapMnsMessage(resp ali_mns.MessageReceiveResponse, topic string) (msg pubsub.NewMessage, err error) {
+	msg.Topic = topic
+	msg.Data = []byte(resp.MessageBody)
+
+	jsonBody, err := json.Marshal(msg)
+	if err != nil {
+		return msg, err
+	}
+
+	json.Unmarshal(jsonBody, &msg.Metadata)
+
+	return msg, nil
+}
+
+func (m *mns) processMessageLoop(ctx context.Context, topic string, queue ali_mns.AliMNSQueue, handler pubsub.Handler, respChan chan ali_mns.MessageReceiveResponse, errChan chan error) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case resp := <-respChan:
+			{
+				m.logger.Infof("response: %+v", resp)
+
+				msg, err := wrapMnsMessage(resp, topic)
+				if err != nil {
+					m.logger.Error(err)
+					continue
+				}
+
+				handler(ctx, &msg)
+
+				m.logger.Debugf("change the visibility: %+v", resp.ReceiptHandle)
+				if ret, e := queue.ChangeMessageVisibility(resp.ReceiptHandle, int64(ReceivingMessageVisibilityTimeout)); e != nil {
+					m.logger.Error(e)
+				} else {
+					m.logger.Infof("visibility changed: %+v", ret)
+					m.logger.Debugf("delete it now: %+v", ret.ReceiptHandle)
+					if e := queue.DeleteMessage(ret.ReceiptHandle); e != nil {
+						m.logger.Error(e)
+					}
+					continue
+				}
+			}
+		case err := <-errChan:
+			{
+				if !ali_mns.ERR_MNS_MESSAGE_NOT_EXIST.IsEqual(err) {
+					m.logger.Error(err)
+				}
+				continue
+			}
+		default:
+			time.Sleep(ReceivingMessageWaitTime)
+		}
+	}
+}
+
+func (m *mns) receiveMessageLoop(ctx context.Context, queue ali_mns.AliMNSQueue, respChan chan ali_mns.MessageReceiveResponse, errChan chan error) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			queue.ReceiveMessage(respChan, errChan, int64(ReceivingMessagePeriod))
+		}
+		time.Sleep(ReceivingMessageWaitTime)
+	}
+}
+
+// Features does nothing here
+func (m *mns) Features() []pubsub.Feature {
+	return nil
+}
+
+// Close unsubscribes all topics/queues and closes this service gracefully
+func (m *mns) Close() error {
+	return nil
```

Comment:
I'm thinking you want a component level `ctx` and `cancel` so you can cleanly stop the goroutines started by the component.